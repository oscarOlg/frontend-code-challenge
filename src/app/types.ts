export interface UserListComponentProps {
    usersData?: UserData;
    hasError: boolean;
    isDataLoading: boolean;
    resetData: () => void;
    handleOnDeleteClick: (id: string) => void;
    handleSearchByCountry: (country: string) => void;
}

export interface FilterBarComponentProps {
    onChangeColor: () => void;
    onToggleCountrySort: () => void;
    resetData: () => void;
    handleSearchByCountry: (country: string) => void;
}

export interface UserData {
    results: User[];
}

export interface User {
    name: Name;
    location: Location; 
    picture: Picture;
    login: Login;
}

export interface Name {
    title: string;
    first: string;
    last: string;
}

export interface Location {
    country: string;
}

export interface Picture {
    large: string;
    medium: string;
    thumbnail: string;
}

export interface Login {
    uuid: string;
}