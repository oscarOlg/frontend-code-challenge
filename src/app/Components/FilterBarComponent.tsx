import React from 'react';

import './styles/FilterBarComponent.css';
import { FilterBarComponentProps } from '../types';

export const FilterBarComponent = (props: FilterBarComponentProps) => {
  return (
    <div data-testid='filter-bar' className='filterbar-container'>
      <button data-testid='change-color-button' onClick={() => props.onChangeColor()}>Color table</button>
      <button data-testid='sort-button' onClick={() => props.onToggleCountrySort()}>Sort by country</button>
      <button data-testid='reset-button' onClick={() => props.resetData()}>reset</button>
      <input data-testid='search-input' placeholder='Search by Country' onChange={e => props.handleSearchByCountry(e.target.value)}></input>
    </div>
  )
}
