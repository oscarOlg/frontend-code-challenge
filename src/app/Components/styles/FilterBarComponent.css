.filterbar-container {
    display: flex;
    justify-content: center;
    gap: 10px;
    height: 30px;
}
