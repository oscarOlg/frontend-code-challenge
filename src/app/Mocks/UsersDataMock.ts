import { UserData } from "../types";

export const userDataMock: UserData = {
  results: [
    {
      name: {
          title: "Mrs",
          first: "Roseni",
          last: "de Souza"
      },
      location: {
          country: "Brazil",

      },
      login: {
          uuid: "abc123",
      },
      picture: {
        large: "https://randomuser.me/api/portraits/women/33.jpg",
        medium: "https://randomuser.me/api/portraits/med/women/33.jpg",
        thumbnail: "https://randomuser.me/api/portraits/thumb/women/33.jpg"
      },
    }
  ]
}