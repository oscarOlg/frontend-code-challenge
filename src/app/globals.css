
* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}

body {
  background: white;
}


button {
  padding: 5px 10px;
  border-radius: 5px;
  color: rgb(255, 255, 255);
  background-color: rgb(0, 101, 148);
}

input {
  border-radius: 5px;
  border-color: rgb(0, 101, 148);
}

input::placeholder {
  font-weight: bold;
  padding: 5px;
  opacity: 0.65;
  color: rgb(0, 101, 148);
}