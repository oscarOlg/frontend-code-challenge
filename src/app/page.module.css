.main {
  display: flex;
  flex-direction: column;
  gap: 30px;
  align-items: center;
  padding: 50px 0;
  min-height: 100vh;
}
