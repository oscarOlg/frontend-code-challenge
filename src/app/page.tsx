"use client";
import styles from './page.module.css'
import { UsersListContainer } from './Components/UsersListContainer';
import { useState, useEffect } from 'react';
import { UserData, UserListComponentProps } from './types';

export const Home = () => {
  const [usersData, setUsersData] = useState<UserData>({results: []});
  const [defaultData, setDefaultData] = useState<UserData>({results: []});
  const [filteredData, setFilteredData] = useState<UserData>({results: []});
  const [isDataLoading, setIsDataLoading] = useState(false); 
  const [error, setError] = useState(false);

  useEffect(() => {
    setFilteredData(usersData)
  }, [usersData])
  

  const resetData = () => {
    setUsersData(defaultData);
    setFilteredData(defaultData);
  }

  const handleOnDeleteClick = (id: string) => {
    const newUsers = filteredData.results.filter(user => user.login.uuid !== id);
    setFilteredData({results: newUsers});
  }

  const handleSearchByCountry = (country: string) => {
    if(country.length) { 
      const normCountry = country.toLocaleLowerCase();
      const filteredData = [...usersData.results].filter(user => user.location.country.toLowerCase().match(normCountry));
      setFilteredData({results: filteredData});
    } else {
      setFilteredData({results: []});
    }
  }

  const UserListProps: UserListComponentProps = {
    usersData: filteredData.results.length ? filteredData : usersData,
    isDataLoading, 
    resetData,
    handleOnDeleteClick,
    handleSearchByCountry,
    hasError: error,
  }

  const fetchUsers = (numOfUsers: number) => {
    fetch(`https://randomuser.me/api/?results=${numOfUsers}`)
      .then(response =>{
        if(response.ok) {
          return response.json();
        }
        throw response;
      })
      .then(data => {
        setUsersData(data);
        setDefaultData(data);
      })
      .catch(error => {
        console.error(error);
        setError(true);
      })
      .finally(() => {
        setIsDataLoading(false);
      })
  }

  useEffect(() => {
    setIsDataLoading(true);
    fetchUsers(100);
  }, []);

  return (
    <main className={styles.main}>
      <h1>Users list</h1>
      {<UsersListContainer {...UserListProps} />} 
    </main>
  )
}
