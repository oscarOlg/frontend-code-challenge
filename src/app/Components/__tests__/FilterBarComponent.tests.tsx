import '@testing-library/jest-dom';
import {render, fireEvent} from '@testing-library/react'
import { FilterBarComponentProps } from '@/app/types';
import { userDataMock } from '../../Mocks/UsersDataMock';
import { FilterBarComponent } from '../FilterBarComponent';


describe('UserListContainer tests', () => { 
    let props: FilterBarComponentProps;

    const renderComponent  = (props: FilterBarComponentProps)  => {
        return render(<FilterBarComponent {...props} />)
    };

    let component: ReturnType<typeof renderComponent>;


    beforeEach(() => {
        props = {
            onChangeColor: () => {},
            onToggleCountrySort: () => {},
            resetData: () => {},
            handleSearchByCountry: (country: string) => {},
        };
    });
      
    afterEach(() => {
        if (component) {
          component.unmount();
        }
    });
    
    
    it('Should render FilterBarComponent correctly', () => {
        component = renderComponent(props);
        const filterBar = component.queryByTestId('filter-bar');
        expect(filterBar).toBeVisible();
        expect(component.baseElement).toMatchSnapshot();
    });

    it('Should handle clicking on change color button', () => {
        props.onChangeColor = jest.fn()
        component = renderComponent(props);
        const changeColor = component.queryByTestId('change-color-button') as HTMLButtonElement;
        expect(changeColor).toBeVisible();
        fireEvent.click(changeColor);
        expect(props.onChangeColor).toHaveBeenCalled();
    });

    it('Should handle clicking sort button', () => {
        props.onToggleCountrySort = jest.fn()
        component = renderComponent(props);
        const sortButton = component.queryByTestId('sort-button') as HTMLButtonElement;
        expect(sortButton).toBeVisible();
        fireEvent.click(sortButton);
        expect(props.onToggleCountrySort).toHaveBeenCalled();
    });

    it('Should handle clicking reset button', () => {
        props.resetData = jest.fn()
        component = renderComponent(props);
        const resetButton = component.queryByTestId('reset-button') as HTMLButtonElement;
        expect(resetButton).toBeVisible();
        fireEvent.click(resetButton);
        expect(props.resetData).toHaveBeenCalled();
    });

    it('Should handle searching by country', () => {
        props.handleSearchByCountry = jest.fn()
        component = renderComponent(props);
        const searchInput = component.queryByTestId('search-input') as HTMLInputElement;
        expect(searchInput).toBeVisible();
        fireEvent.change(searchInput, {target: {value: 'Mexico'}});
        expect(props.handleSearchByCountry).toHaveBeenCalled();
    });

});