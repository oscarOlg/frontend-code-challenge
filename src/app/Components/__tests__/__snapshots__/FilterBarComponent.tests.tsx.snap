// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListContainer tests Should render FilterBarComponent correctly 1`] = `
<body>
  <div>
    <div
      class="filterbar-container"
      data-testid="filter-bar"
    >
      <button
        data-testid="change-color-button"
      >
        Color table
      </button>
      <button
        data-testid="sort-button"
      >
        Sort by country
      </button>
      <button
        data-testid="reset-button"
      >
        reset
      </button>
      <input
        data-testid="search-input"
        placeholder="Search by Country"
      />
    </div>
  </div>
</body>
`;
